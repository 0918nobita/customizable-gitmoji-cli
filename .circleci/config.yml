version: 2.1

executors:
  node-lts:
    docker:
      - image: circleci/node:lts

jobs:
  install:
    executor: node-lts
    steps:
      - checkout
      - restore_cache:
          keys:
            - dependency-cache-{{ checksum "package-lock.json" }}
            - dependency-cache-
      - run: npm ci
      - save_cache:
          key: dependency-cache-{{ checksum "package-lock.json" }}
          paths:
            - node_modules
  test:
    executor: node-lts
    steps:
      - checkout
      - run: npm test

workflows:
  version: 2
  test:
    jobs:
      - install
      - test
