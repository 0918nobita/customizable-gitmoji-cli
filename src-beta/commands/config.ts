import Conf from 'conf'
import inquirer from 'inquirer'

// eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
inquirer.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'))

const config = new Conf()

export enum ConfigKeys {
  AUTO_ADD = 'auto_add',
  EMOJI_FORMAT = 'emoji_format',
  SIGNED_COMMIT = 'signed_commit',
  TITLE_MAX_LENGTH = 'title_max_length'
}

type Config = {
  [ConfigKeys.AUTO_ADD]: boolean
  [ConfigKeys.EMOJI_FORMAT]: 'emoji' | 'code'
  [ConfigKeys.SIGNED_COMMIT]: boolean
  [ConfigKeys.TITLE_MAX_LENGTH]: number
}

export function getConfig(): Config {
  return {
    [ConfigKeys.AUTO_ADD]: config.get(ConfigKeys.AUTO_ADD, false),
    [ConfigKeys.EMOJI_FORMAT]: config.get(ConfigKeys.EMOJI_FORMAT, 'code'),
    [ConfigKeys.SIGNED_COMMIT]: config.get(ConfigKeys.SIGNED_COMMIT, false),
    [ConfigKeys.TITLE_MAX_LENGTH]: config.get(ConfigKeys.TITLE_MAX_LENGTH, 48)
  }
}

export default async function() {
  const answer = await inquirer.prompt([
    {
      name: ConfigKeys.AUTO_ADD,
      message: 'Enable automatic "git add ."',
      type: 'confirm'
    },
    {
      name: ConfigKeys.EMOJI_FORMAT,
      message: 'Select how emojis should be used in commits',
      type: 'list',
      choices: [
        { name: ':smile:', value: 'code' },
        { name: 'ğŸ˜„', value: 'emoji' }
      ]
    },
    {
      name: ConfigKeys.SIGNED_COMMIT,
      message: 'Enable signed commits',
      type: 'confirm'
    },
    {
      name: ConfigKeys.TITLE_MAX_LENGTH,
      message: 'Max length of commit title',
      type: 'number',
      validate: number => number > 0 && number <= 72,
      default: 48
    }
  ])
  config.set(ConfigKeys.AUTO_ADD, answer[ConfigKeys.AUTO_ADD])
  config.set(ConfigKeys.EMOJI_FORMAT, answer[ConfigKeys.EMOJI_FORMAT])
  config.set(ConfigKeys.SIGNED_COMMIT, answer[ConfigKeys.SIGNED_COMMIT])
  config.set(ConfigKeys.TITLE_MAX_LENGTH, answer[ConfigKeys.TITLE_MAX_LENGTH])
}
