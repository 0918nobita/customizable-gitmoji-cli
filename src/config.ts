import cosmiconfig from 'cosmiconfig'

const configExplorer = cosmiconfig('gitmoji', { cache: false })

type Config = {
  autoAdd: boolean
  emojiFormat: 'emoji' | 'code'
  signedCommit: boolean
  titleMaxLength: number
  presets: string[] | string | null
  rules: Gitmoji[]
  scopes: []
  order: string[]
}

function validateGitmoji(value: unknown): value is Gitmoji {
  if (typeof value === 'object' && value) {
    if (!('emoji' in value && 'name' in value && 'description' in value))
      return false
    if (typeof (value as { emoji: unknown }).emoji !== 'string') return false
    if (typeof (value as { name: unknown }).name !== 'string') return false
    if (typeof (value as { description: unknown }).description !== 'string')
      return false
    return true
  }
  return false
}

function validate<T extends keyof Config>(
  key: T,
  value: unknown
): value is Config[T] {
  // eslint-disable-next-line default-case
  switch (key) {
    case 'autoAdd':
      return typeof value === 'boolean'
    case 'emojiFormat':
      return (
        typeof value === 'string' && (value === 'emoji' || value === 'code')
      )
    case 'signedCommit':
      return typeof value === 'boolean'
    case 'titleMaxLength':
      return typeof value === 'number' && value > 0 && value <= 72
    case 'presets':
      return (
        typeof value === 'string' ||
        (Array.isArray(value) && value.every(v => typeof v === 'string'))
      )
    case 'rules':
      return Array.isArray(value) && value.every(v => validateGitmoji(v))
    case 'order':
      return Array.isArray(value) && value.every(v => typeof v === 'string')
    case 'scopes':
      return Array.isArray(value) && value.every(v => typeof v === 'string')
  }
  throw new Error(`The given config ${key} does not meet the conditions.`)
}

const defaultConfig: Config = {
  autoAdd: false,
  emojiFormat: 'code',
  signedCommit: false,
  titleMaxLength: 48,
  order: [],
  rules: [],
  scopes: [],
  presets: 'base'
}

let cache: Config | undefined

export async function saveConfig(searchFrom?: string) {
  const result = await configExplorer.search(searchFrom)
  if (result && !result.isEmpty) {
    if ('rules' in result.config && validate('rules', result.config.rules))
      defaultConfig.presets = null

    const {
      autoAdd,
      emojiFormat,
      signedCommit,
      titleMaxLength,
      rules,
      order,
      presets,
      scopes
    } = Object.assign(defaultConfig, result.config)

    validate('autoAdd', autoAdd)
    validate('emojiFormat', emojiFormat)
    validate('signedCommit', signedCommit)
    validate('titleMaxLength', titleMaxLength)
    validate('order', order)
    validate('presets', presets)
    validate('scopes', scopes)

    cache = {
      autoAdd,
      emojiFormat,
      signedCommit,
      titleMaxLength,
      rules,
      order,
      presets,
      scopes
    }
    return cache
  }
  return defaultConfig
}

export default function() {
  if (!cache) throw new Error('Something wrong.')
  return cache
}
