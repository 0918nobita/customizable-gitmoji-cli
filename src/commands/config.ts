import consola from 'consola'
import chalk from 'chalk'
import inquirer from 'inquirer'
import { ConfigKeys, setConfig, defaults, getConfigs } from '../config'

// eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
inquirer.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'))

export async function list() {
  const configs = await getConfigs()
  const longest = Object.keys(configs)
    .map(c => c.length)
    .reduce((accumulator, current) => Math.max(accumulator, current))
  Object.entries(configs).forEach(({ 0: key, 1: value }) => {
    consola.info(
      [
        key,
        ''.padStart(longest - key.length + 1, ' '),
        ': ',
        chalk.yellow(String(value))
      ].join('')
    )
  })
}

export default async function() {
  const answer = await inquirer.prompt([
    {
      name: ConfigKeys.AUTO_ADD,
      message: 'Enable automatic "git add ."',
      type: 'confirm',
      default: defaults[ConfigKeys.AUTO_ADD]
    },
    {
      name: ConfigKeys.EMOJI_FORMAT,
      message: 'Select how emojis should be used in commits',
      type: 'list',
      choices: [
        { name: ':smile:', value: 'code' },
        { name: 'ğŸ˜„', value: 'emoji' }
      ],
      default: defaults[ConfigKeys.EMOJI_FORMAT]
    },
    {
      name: ConfigKeys.SIGNED_COMMIT,
      message: 'Enable signed commits',
      type: 'confirm',
      default: defaults[ConfigKeys.SIGNED_COMMIT]
    },
    {
      name: ConfigKeys.TITLE_MAX_LENGTH,
      message: 'Max length of commit title',
      type: 'number',
      validate: number => number > 0 && number <= 72,
      default: defaults[ConfigKeys.TITLE_MAX_LENGTH]
    }
  ])
  setConfig(ConfigKeys.AUTO_ADD, answer[ConfigKeys.AUTO_ADD])
  setConfig(ConfigKeys.EMOJI_FORMAT, answer[ConfigKeys.EMOJI_FORMAT])
  setConfig(ConfigKeys.SIGNED_COMMIT, answer[ConfigKeys.SIGNED_COMMIT])
  setConfig(ConfigKeys.TITLE_MAX_LENGTH, answer[ConfigKeys.TITLE_MAX_LENGTH])
}
