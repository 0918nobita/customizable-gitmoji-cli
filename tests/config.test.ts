import { getConfig, setConfig, ConfigKeys } from '~/config'

describe('config module', () => {
  it('should match for AUTO ADD', () => {
    setConfig(ConfigKeys.AUTO_ADD, true)
    expect(getConfig(ConfigKeys.AUTO_ADD)).toMatchSnapshot()
  })

  it('should match for EMOJI FORMAT', () => {
    setConfig(ConfigKeys.EMOJI_FORMAT, 'emoji')
    expect(getConfig(ConfigKeys.EMOJI_FORMAT)).toMatchSnapshot()
  })

  it('should match for SIGNED COMMIT', () => {
    setConfig(ConfigKeys.SIGNED_COMMIT, false)
    expect(getConfig(ConfigKeys.SIGNED_COMMIT)).toMatchSnapshot()
  })

  it('should match for TITLE MAX LENGTH', () => {
    setConfig(ConfigKeys.TITLE_MAX_LENGTH, 48)
    expect(getConfig(ConfigKeys.TITLE_MAX_LENGTH)).toMatchSnapshot()
  })
})
