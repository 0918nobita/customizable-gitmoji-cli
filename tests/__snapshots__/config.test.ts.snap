// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config module should match for AUTO ADD 1`] = `true`;

exports[`config module should match for EMOJI FORMAT 1`] = `"emoji"`;

exports[`config module should match for SIGNED COMMIT 1`] = `false`;

exports[`config module should match for TITLE MAX LENGTH 1`] = `48`;
